File: go/gate.go
package koosgate

import "time"

type GateResult string

type ExecutionStatus string

type EpistemicStatus string

const (
	Allow       GateResult = "ALLOW"
	Deny        GateResult = "DENY"
	HumanReview GateResult = "HUMAN_REVIEW"

	NotAttempted ExecutionStatus = "NOT_ATTEMPTED"
	Succeeded    ExecutionStatus = "SUCCEEDED"
	Failed       ExecutionStatus = "FAILED"

	Observed     EpistemicStatus = "OBSERVED"
	Derived      EpistemicStatus = "DERIVED"
	Claimed      EpistemicStatus = "CLAIMED"
	Inferred     EpistemicStatus = "INFERRED"
	Verified     EpistemicStatus = "VERIFIED"
	Unknown      EpistemicStatus = "UNKNOWN"
	Contradicted EpistemicStatus = "CONTRADICTED"
)

type EvidenceResult struct {
	Subject         string
	Claims          map[string]string
	Source          string
	ObservedAt      time.Time
	IssuedAt        time.Time
	ExpiresAt       time.Time
	Validity        bool
	Provenance      string
	EpistemicStatus EpistemicStatus
}

func (e EvidenceResult) IsTemporallyValid(at time.Time) bool {
	return !at.Before(e.IssuedAt) && !at.After(e.ExpiresAt)
}

type AuthorityLease struct {
	Subject     string
	Scope       string
	Purpose     string
	Constraints map[string]string
	IssuedAt    time.Time
	ExpiresAt   time.Time
	Issuer      string
}

func (a AuthorityLease) Matches(subject, scope string) bool {
	return a.Subject == subject && a.Scope == scope
}

func (a AuthorityLease) IsTemporallyActive(at time.Time) bool {
	return !at.Before(a.IssuedAt) && !at.After(a.ExpiresAt)
}

func (a AuthorityLease) SatisfiesConstraints(required map[string]string) bool {
	for key, value := range required {
		if a.Constraints[key] != value {
			return false
		}
	}
	return true
}

func (a AuthorityLease) IsValid(at time.Time, subject, scope string) bool {
	return a.Matches(subject, scope) && a.IsTemporallyActive(at)
}

type Actor struct {
	ID           string
	Capabilities map[string]bool
	Authority    []AuthorityLease
}

type GateRequest struct {
	Subject             string
	Capability          string
	Scope               string
	RequiredConstraints map[string]string
	PolicySatisfied     bool
	PolicySpecified     bool
	ConflictMode        GateResult
}

type PackCharacter struct {
	Name         string
	Role         string
	Function     string
	Contribution string
}

var PackCharacters = []PackCharacter{
	{"QUILL", "Overseer/Coherence", "Coordinates constitutional consistency", "Checks whole-system coherence before decision"},
	{"RUNE", "Systems Architect/Structure", "Shapes interfaces, contracts, and boundaries", "Keeps the gate contract structurally sound"},
	{"ZYRA", "Crypto/Innovation", "Explores stronger technical evidence mechanisms", "Improves evidence quality without becoming authority"},
	{"VERA", "Feasibility/Reality Check", "Tests viability against reality", "Challenges impractical assumptions in decisions"},
	{"MAVEN", "Knowledge/Curation", "Curates context and retrieval", "Supplies knowledge inputs without minting authority"},
	{"ARIA", "Meaning/Expression", "Turns system state into human-facing explanation", "Makes outcomes understandable"},
	{"LYRA", "Human Intelligence/Context", "Maintains human-centered reasoning", "Protects usability and operator clarity"},
	{"SOL", "Operations/Momentum", "Coordinates approved operational flow", "Moves allowed work toward runtime without granting permission"},
	{"TALON", "Security/Adversarial", "Assumes hostile conditions and ambiguity", "Tests boundary failures and abuse cases"},
	{"KAI", "Engineering/Implementation", "Builds the actual systems", "Turns contracts into executable implementations"},
	{"SAGE", "Learning/Evaluation", "Evaluates patterns and feedback", "Improves models while respecting gate boundaries"},
	{"NARA", "Strategy/Synthesis", "Connects present decisions to long-horizon outcomes", "Frames system trade-offs and opportunity"},
	{"NYX", "Boundary/Anomaly", "Represents unknowns and edge cases", "Surfaces exceptional states for careful handling"},
}

type GateDecision struct {
	Decision         GateResult
	Predicates       map[string]bool
	Reasons          []string
	EvidenceConflict bool
}

type RuntimeEvent struct {
	Action string
	Status ExecutionStatus
	Reason string
}

type LedgerEntry struct {
	EntryType  string
	RecordedAt time.Time
	Payload    map[string]string
}

type Ledger struct {
	Entries []LedgerEntry
}

func (l *Ledger) Append(entryType string, payload map[string]string, recordedAt time.Time) {
	copyPayload := map[string]string{}
	for key, value := range payload {
		copyPayload[key] = value
	}
	l.Entries = append(l.Entries, LedgerEntry{EntryType: entryType, RecordedAt: recordedAt, Payload: copyPayload})
}

type Gate struct{}

func (r GateRequest) policySatisfied() bool {
	if !r.PolicySpecified {
		return true
	}
	return r.PolicySatisfied
}

func (r GateRequest) conflictMode() GateResult {
	if r.ConflictMode == "" {
		return Deny
	}
	return r.ConflictMode
}

func (Gate) Evaluate(actor Actor, request GateRequest, evidence []EvidenceResult, at time.Time) GateDecision {
	matching := []AuthorityLease{}
	active := []AuthorityLease{}
	for _, lease := range actor.Authority {
		if lease.Matches(request.Subject, request.Scope) {
			matching = append(matching, lease)
			if lease.IsValid(at, request.Subject, request.Scope) {
				active = append(active, lease)
			}
		}
	}
	capabilitySatisfied := actor.Capabilities[request.Capability]
	policySatisfied := request.policySatisfied()
	authorityValid := len(active) > 0
	subjectValid := actor.ID == request.Subject
	temporalValid := len(evidence) > 0 && len(active) > 0
	evidenceConflict := false
	evidenceValid := len(evidence) > 0
	firstValidity := false
	haveValidity := false
	for _, item := range evidence {
		if item.Subject != request.Subject {
			subjectValid = false
		}
		if !item.IsTemporallyValid(at) {
			temporalValid = false
		}
		if item.EpistemicStatus == Contradicted {
			evidenceConflict = true
		}
		if !item.Validity {
			evidenceValid = false
		}
		if !haveValidity {
			firstValidity = item.Validity
			haveValidity = true
		} else if firstValidity != item.Validity {
			evidenceConflict = true
		}
	}
	constraintsSatisfied := true
	if len(request.RequiredConstraints) > 0 {
		constraintsSatisfied = false
		for _, lease := range active {
			if lease.SatisfiesConstraints(request.RequiredConstraints) {
				constraintsSatisfied = true
				break
			}
		}
	}
	evidenceValid = evidenceValid && !evidenceConflict && len(evidence) > 0
	predicates := map[string]bool{
		"C": capabilitySatisfied,
		"P": policySatisfied,
		"A": authorityValid,
		"T": temporalValid,
		"S": subjectValid,
		"R": constraintsSatisfied,
		"E": evidenceValid,
	}
	reasons := []string{}
	for _, key := range []string{"C", "P", "A", "T", "S", "R", "E"} {
		if !predicates[key] {
			reasons = append(reasons, key)
		}
	}
	decision := Deny
	if evidenceConflict {
		decision = request.conflictMode()
	} else if capabilitySatisfied && policySatisfied && authorityValid && temporalValid && subjectValid && constraintsSatisfied && evidenceValid {
		decision = Allow
	}
	return GateDecision{Decision: decision, Predicates: predicates, Reasons: reasons, EvidenceConflict: evidenceConflict}
}

type Runtime struct{}

func (Runtime) Execute(decision GateDecision, action string, shouldSucceed bool) RuntimeEvent {
	if decision.Decision != Allow {
		return RuntimeEvent{Action: action, Status: NotAttempted, Reason: "Gate denied execution"}
	}
	if shouldSucceed {
		return RuntimeEvent{Action: action, Status: Succeeded, Reason: "Runtime completed action"}
	}
	return RuntimeEvent{Action: action, Status: Failed, Reason: "Runtime failed after gate approval"}
}

File: typescript/gate.ts
export type GateResult = 'ALLOW' | 'DENY' | 'HUMAN_REVIEW';
export type ExecutionStatus = 'NOT_ATTEMPTED' | 'SUCCEEDED' | 'FAILED';
export type EpistemicStatus = 'OBSERVED' | 'DERIVED' | 'CLAIMED' | 'INFERRED' | 'VERIFIED' | 'UNKNOWN' | 'CONTRADICTED';

export interface EvidenceResult {
  subject: string;
  claims: Record<string, unknown>;
  source: string;
  observed_at: Date;
  issued_at: Date;
  expires_at: Date;
  validity: boolean;
  provenance: string;
  epistemic_status: EpistemicStatus;
}

export interface EvidenceAdapter<T> {
  toEvidence(raw: T): EvidenceResult;
}

export class ObjectEvidenceAdapter implements EvidenceAdapter<Record<string, unknown>> {
  private readonly sourceName: string;

  constructor(sourceName: string) {
    this.sourceName = sourceName;
  }

  toEvidence(raw: Record<string, unknown>): EvidenceResult {
    return {
      subject: String(raw.subject),
      claims: (raw.claims as Record<string, unknown>) ?? {},
      source: this.sourceName,
      observed_at: raw.observed_at as Date,
      issued_at: raw.issued_at as Date,
      expires_at: raw.expires_at as Date,
      validity: Boolean(raw.validity),
      provenance: String(raw.provenance ?? this.sourceName),
      epistemic_status: (raw.epistemic_status as EpistemicStatus) ?? 'UNKNOWN',
    };
  }
}

export interface AuthorityLease {
  subject: string;
  scope: string;
  purpose: string;
  constraints: Record<string, string>;
  issued_at: Date;
  expires_at: Date;
  issuer: string;
}

export interface Actor {
  id: string;
  capabilities: string[];
  authority: AuthorityLease[];
}

export interface GateRequest {
  subject: string;
  capability: string;
  scope: string;
  required_constraints?: Record<string, string>;
  policy_satisfied?: boolean;
  conflict_mode?: GateResult;
}

export interface PackCharacter {
  name: string;
  role: string;
  function: string;
  contribution: string;
}

export const PACK_CHARACTERS: PackCharacter[] = [
  { name: 'QUILL', role: 'Overseer/Coherence', function: 'Coordinates constitutional consistency', contribution: 'Checks whole-system coherence before decision' },
  { name: 'RUNE', role: 'Systems Architect/Structure', function: 'Shapes interfaces, contracts, and boundaries', contribution: 'Keeps the gate contract structurally sound' },
  { name: 'ZYRA', role: 'Crypto/Innovation', function: 'Explores stronger technical evidence mechanisms', contribution: 'Improves evidence quality without becoming authority' },
  { name: 'VERA', role: 'Feasibility/Reality Check', function: 'Tests viability against reality', contribution: 'Challenges impractical assumptions in decisions' },
  { name: 'MAVEN', role: 'Knowledge/Curation', function: 'Curates context and retrieval', contribution: 'Supplies knowledge inputs without minting authority' },
  { name: 'ARIA', role: 'Meaning/Expression', function: 'Turns system state into human-facing explanation', contribution: 'Makes outcomes understandable' },
  { name: 'LYRA', role: 'Human Intelligence/Context', function: 'Maintains human-centered reasoning', contribution: 'Protects usability and operator clarity' },
  { name: 'SOL', role: 'Operations/Momentum', function: 'Coordinates approved operational flow', contribution: 'Moves allowed work toward runtime without granting permission' },
  { name: 'TALON', role: 'Security/Adversarial', function: 'Assumes hostile conditions and ambiguity', contribution: 'Tests boundary failures and abuse cases' },
  { name: 'KAI', role: 'Engineering/Implementation', function: 'Builds the actual systems', contribution: 'Turns contracts into executable implementations' },
  { name: 'SAGE', role: 'Learning/Evaluation', function: 'Evaluates patterns and feedback', contribution: 'Improves models while respecting gate boundaries' },
  { name: 'NARA', role: 'Strategy/Synthesis', function: 'Connects present decisions to long-horizon outcomes', contribution: 'Frames system trade-offs and opportunity' },
  { name: 'NYX', role: 'Boundary/Anomaly', function: 'Represents unknowns and edge cases', contribution: 'Surfaces exceptional states for careful handling' },
];

export interface GateDecision {
  decision: GateResult;
  predicates: Record<string, boolean>;
  reasons: string[];
  evidence_conflict: boolean;
}

export interface RuntimeEvent {
  action: string;
  status: ExecutionStatus;
  reason: string;
}

export interface LedgerEntry {
  entry_type: 'decision' | 'event';
  recorded_at: Date;
  payload: Record<string, string>;
}

export class Ledger {
  private readonly history: LedgerEntry[] = [];

  append(entry: LedgerEntry): void {
    this.history.push({ ...entry, payload: { ...entry.payload } });
  }

  entries(): LedgerEntry[] {
    return [...this.history];
  }
}

export function leaseIsValid(lease: AuthorityLease, at: Date, subject: string, scope: string): boolean {
  return lease.subject === subject && lease.scope === scope && lease.issued_at <= at && at <= lease.expires_at;
}

export function gate(actor: Actor, request: GateRequest, evidence: EvidenceResult[], at: Date): GateDecision {
  const requiredConstraints = request.required_constraints ?? {};
  const activeLeases = actor.authority.filter((lease) => leaseIsValid(lease, at, request.subject, request.scope));
  const capability = actor.capabilities.includes(request.capability);
  const policy = request.policy_satisfied ?? true;
  const authority = activeLeases.length > 0;
  const subject = actor.id === request.subject && evidence.every((item) => item.subject === request.subject);
  const temporal = evidence.length > 0 && activeLeases.length > 0 && evidence.every((item) => item.issued_at <= at && at <= item.expires_at);
  const constraints = Object.keys(requiredConstraints).length === 0
    ? true
    : activeLeases.some((lease) => Object.entries(requiredConstraints).every(([key, value]) => lease.constraints[key] === value));
  const evidenceConflict = evidence.some((item) => item.epistemic_status === 'CONTRADICTED')
    || new Set(evidence.map((item) => item.validity)).size > 1;
  const evidenceValid = evidence.length > 0 && evidence.every((item) => item.validity) && !evidenceConflict;
  const predicates = { C: capability, P: policy, A: authority, T: temporal, S: subject, R: constraints, E: evidenceValid };
  return {
    decision: evidenceConflict ? (request.conflict_mode ?? 'DENY') : (Object.values(predicates).every(Boolean) ? 'ALLOW' : 'DENY'),
    predicates,
    reasons: Object.entries(predicates).filter(([, ok]) => !ok).map(([key]) => key),
    evidence_conflict: evidenceConflict,
  };
}

export function runtime(decision: GateDecision, action: string, shouldSucceed: boolean): RuntimeEvent {
  if (decision.decision !== 'ALLOW') {
    return { action, status: 'NOT_ATTEMPTED', reason: 'Gate denied execution' };
  }
  return shouldSucceed
    ? { action, status: 'SUCCEEDED', reason: 'Runtime completed action' }
    : { action, status: 'FAILED', reason: 'Runtime failed after gate approval' };
}
